### PR TITLE
feat(condo): DOMA-7499 schema MarketItemFile

### PR DESCRIPTION
**Pull request by VKislov**

### CODE REVIEW

File: apps/condo/domains/marketplace/schema/MarketItemFile.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                condition: 'Q(deletedAt__isnull=True)',
+                name: 'MarketItemFile_unique_organization_marketItem',
+            },
+        ],
+    },
+    hooks: {
+        afterChange: fileMetaAfterChange,
+        // afterDelete: async ({ existingItem }) => {
+        //     await Adapter.delete(existingItem.file)
+        // },
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/marketplace/schema/MarketItemFile.js:78]: remove comment

File: apps/condo/domains/marketplace/schema/MarketItemFile.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+]
+const ERRORS = {
+    NON_IMAGE_FILE: {
+        code: BAD_USER_INPUT,
+        type: 'NON_IMAGE_FILE',
+    },
+}
+
+const MarketItemFile = new GQLListSchema('MarketItemFile', {
+    schemaDoc: 'File attached to the market item',
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/marketplace/schema/MarketItemFile.js:33]: What is this file?
Will it be only for the image of the product or could it be some additional files (such as instructions)?

File: apps/condo/domains/marketplace/access/MarketItemFile.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    return false
+}
+
+/*
+  Rules are logical functions that used for list access, and may return a boolean (meaning
+  all or no items are available) or a set of filters that limit the available items.
+*/
+module.exports = {
+    canReadMarketItemFiles,
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/access/MarketItemFile.js:55]:
```suggestion
    canReadMarketItemFiles: canReadMarketItems,
```

File: apps/condo/domains/marketplace/access/MarketItemFile.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+const { throwAuthenticationError } = require('@open-condo/keystone/apolloErrorFormatter')
+const { getById } = require('@open-condo/keystone/schema')
+
+const { canReadMarketItems } = require('@condo/domains/marketplace/access/MarketItem')
+const { checkPermissionInUserOrganizationOrRelatedOrganization } = require('@condo/domains/organization/utils/accessSchema')
+const { RESIDENT } = require('@condo/domains/user/constants/common')
+const { STAFF } = require('@condo/domains/user/constants/common')
+
+const canReadMarketItemFiles = canReadMarketItems
+
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/access/MarketItemFile.js:16]:
```suggestion
```

File: apps/condo/domains/marketplace/access/MarketItemFile.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    if (user.type === RESIDENT) return false
+
+    if (user.type === STAFF) {
+        if (operation === 'create') {
+            const marketItemId = get(originalInput, ['marketItem', 'connect', 'id'], null)
+
+            if (marketItemId) {
+                const marketItem = await getById('MarketItem', marketItemId)
+                const organizationId = get(marketItem, 'organization', null)
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/access/MarketItemFile.js:30]:
```suggestion
                const organizationId = get(marketItem, 'organization', null)
                
                if (!organizationId) return false
```

File: apps/condo/domains/marketplace/access/MarketItemFile.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            }
+
+            return false
+        }
+
+        const marketItemFile = await getById('MarketItemFile', itemId)
+        if (!marketItemFile) return false
+
+        const { createdBy, organization } = marketItemFile
+        if (!organization) return createdBy === user.id
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/access/MarketItemFile.js:42]: Everyone with permission should edit files.
```suggestion
        const {organization } = marketItemFile
        if (!organization) return false
```

File: apps/condo/domains/marketplace/components/.gitkeep
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/components/.gitkeep:1]: Please remove this file

File: apps/condo/domains/marketplace/schema/MarketItemFile.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+const ALLOWED_MIME_TYPES = [
+    'image/jpeg',
+    'image/png',
+    'image/heic',
+]
+const ERRORS = {
+    NON_IMAGE_FILE: {
+        code: BAD_USER_INPUT,
+        type: 'NON_IMAGE_FILE',
+    },
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/schema/MarketItemFile.js:29]: Try something like this to keep all error fields together.
```suggestion
    FORBIDDEN_FILE_TYPE: (mimeType) => ({
        code: BAD_USER_INPUT,
        type: 'NON_IMAGE_FILE',
        message: `Expected file to be one of the following mimetypes: ${ALLOWED_MIME_TYPES.map(type => `"${type}"`).join(', ')}. But got: ${mimetype}`,
    }),
```

File: apps/condo/domains/marketplace/schema/MarketItemFile.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            adapter: Adapter,
+            isRequired: true,
+            hooks: {
+                validateInput: ({ resolvedData, fieldPath, context }) => {
+                    const mimetype = get(resolvedData, [fieldPath, 'mimetype'])
+                    if (!ALLOWED_MIME_TYPES.includes(mimetype)) {
+                        throw new GQLError({
+                            ...ERRORS.NON_IMAGE_FILE,
+                            message: `Expected file to be one of the following mimetypes: ${ALLOWED_MIME_TYPES.map(type => `"${type}"`).join(', ')}. But got: ${mimetype}`,
+                        }, context)
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/schema/MarketItemFile.js:57]:
```suggestion
                        throw new GQLError(ERRORS.FORBIDDEN_FILE_TYPE(mimetype), context)
```

File: apps/condo/domains/marketplace/schema/MarketItemFile.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                    }
+                },
+            },
+        },
+
+    },
+    kmigratorOptions: {
+        constraints: [
+            {
+                type: 'models.UniqueConstraint',
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/schema/MarketItemFile.js:67]: What about two images for the same market item?

File: apps/condo/domains/marketplace/schema/MarketItemFile.test.js [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+                    expect.objectContaining({
+                        id: obj.id,
+                    }),
+                ]))
+            })
+        })
+    })
+
+    describe('constraint tests', () => {
+        test('can\'t create two MarketItemFile for one MarketItem in one organization', async () => {
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/schema/MarketItemFile.test.js:296]: We can create an image gallery for market items according to the design.

File: apps/condo/domains/marketplace/access/MarketItemFile.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            return false
+        }
+
+        const marketItemFile = await getById('MarketItemFile', itemId)
+        if (!marketItemFile) return false
+
+        const { organization } = marketItemFile
+        if (!organization) return false
+
+        return await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organization, 'canManageMarketItems')
```
Comment (AleX83Xpert) [inline apps/condo/domains/marketplace/access/MarketItemFile.js:43]: Please refactor logic to call checking of permissions only once. Maybe something like the following:
```suggestion
        let marketItem
        if (operation === 'create') {
            const marketItemId = get(originalInput, ['marketItem', 'connect', 'id'], null)
            if (marketItemId) {
               marketItem = await getById('MarketItem', marketItemId)
            }
        } else if (operation === 'update') {
            const marketItemFile = await getById('MarketItemFile', itemId)
            if (marketItemFile) {
                marketItem = await getById('MarketItem', marketItemFile.marketItem)
            }
        }
        
        organizationId = get(marketItem, 'organization', null)
        if (!organizationId) return false

        return await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canManageMarketItems')
```

**Review by AleX83Xpert — APPROVED** [review]

**Pull request merged**